seed/seedwriter/writer.go

```diff
@@ -865,7 +876,95 @@ func (w *Writer) checkBase(info *snap.Info, modes []string) error {
 	return w.policy.checkBase(info, modes, w.availableByMode)
 }
 
-func (w *Writer) downloaded(seedSnaps []*SeedSnap) error {
+func isKernelSnap(sn *SeedSnap) bool {
+	return sn.modelSnap != nil && sn.modelSnap.SnapType == "kernel"
+}
+
+// snapdCarryingSnapsKnown returns true once it all the snaps that carry snapd
+// or parts of it have been captured; they need to be known before assertions
+// can be fetched with the correct max formats
+func (w *Writer) snapdCarryingSnapsKnown() bool {
+	return w.systemSnap != nil && (w.noKernelSnap || w.kernelSnap != nil)
+}
+
+func (w *Writer) considerForSnapdCarrying(sn *SeedSnap) error {
+	if w.systemSnap == nil {
+		if w.policy.isSystemSnapCandidate(sn) {
+			w.systemSnap = sn
+			return nil
+		}
+	}
+	if w.noKernelSnap {
+		return nil
+	}
+	if isKernelSnap(sn) {
+		w.kernelSnap = sn
+		return nil
+	}
+	w.noKernelSnap = true
+	return nil
+}
+
+// An AssertsFetchFunc should fetch appropriate assertions for the snap sn, it
+// can take into account format constraints caused by the given systemSnap and
+// kernelSnap if set. The returned references are expected to be resolvable
+// in the writing assertion database.
+type AssertsFetchFunc func(sn, systemsSnap, kernelSnap *SeedSnap) ([]*asserts.Ref, error)
+
+func (w *Writer) ensureARefs(upToSnap *SeedSnap, fetchAsserts AssertsFetchFunc) error {
+	n := len(w.snapsFromModel)
+	xn := len(w.extraSnaps)
+
+	iterateUpTo := func(count *int, f func(sn *SeedSnap) error) error {
```

Comment:
It took me a little while to figure this one out; A few suggestions:
I'd rather this function is called something else, because it does not just iterate up to something. It executes a function on a range of SeedSnaps. 
The second is, instead of taking a pointer to a modifiable index, I'd rather we provide it with the startIndex, and let the function return the index of where `upToSnap` was, this is more explicit and easier to understand. This also tells us that actually `count` is an index and not a count. This also has the benefit that if we encounter an error during execution of the callbacks, we don't leave a partially modified `count` variable.

My suggestion for this function prototype would be something like:
```suggestion
        // invokes 'f' on each element in the combined lists of 'w.snapsFromModel' and 'w.extraSnaps',
        // either till end of list, or untill 'upToSnap' is found in one of the lists. It then returns that index.
	invokeRange := func(startIndex int, f func(sn *SeedSnap) error) (error, int) {
```

Reply:
I can do this, I don't like invokeRange though, I can turn it into applyOverRangeUpTo.  Reminder about the return args, error always goes last

Reply:
I made the change, tbh the version without the pointer is maybe a bit more understandable, not sure, but it's also a bit messier/easier to get wrong (I got it wrong the first time I tried to do the change)

Reply:
`applyFromUpTo` works for me too!